### PR TITLE
Implement graceful exit, fix memory leaks & incorrect memory deallocation

### PR DESCRIPTION
**Pull request by and3rson**
This PR introduces several improvements:
- `exit()`/`atexit()` is replaced with `setjmp()`/`longjmp()` combo to allow graceful shutdown with cleanup handler. In a nutshell, `exit()` is replaced with `main_exit()` which performs `longjmp`. So now Nofrendo can be exited with `event_quit`.
- In `nes/nes_rom.c`, memory for `rom` and `vrom` was allocated WITHOUT using guards (`mem_alloc`) but freed WITH using guards (`NOFRENDO_FREE`). This caused the "mem_deleteblock ... -- block not found" which called `exit()` and thus crashed the app. This is now fixed.
- Memory corruption in `bitmap.c` is still present, however it doesn't seem to cause any issues. So I replaced `ASSERT_MSG` with a log message to treat corrupted memory blocks as non-critical issue and proceed with a normal shutdown.

All changes tested on actual hardware: I was able to successfully exit the emulator back into my main firmware with no memory leaks.

### CODE REVIEW

**Comment by moononournation** [comment]
This is a Arduino library repository. If you change the API, please also update the example.

**Comment by moononournation** [comment]
why sometimes change NOFRENDO_FREE() to free()?

**Comment by and3rson** [comment]
@moononournation
- There's no changes to API, the example will still work and run first ROM again after exit. This PR simply prevents it from crashing when exiting. :)
- `NOFRENDO_FREE()` was changed to `free()` because memory for those variables was allocated using `malloc`, not `NOFRENDO_MALLOC`. So it must be freed with `free()` or memory leak will occur. This also fixes "block corrupt" issues like #2 and #9.

In case of my PR, [`*(rominfo)->rom` requires `free()`](https://github.com/moononournation/arduino-nofrendo/pull/18/files#diff-ff9793478d4ac152f6315028a9785af2cf4b839e7d29b168c71dbe00e1fae844R510) instead of `NOFRENDO_FREE()` because [it was allocated using `malloc()`](https://github.com/and3rson/arduino-nofrendo/blob/a356b247b5775ff07410bcd9667b676a79bcda7d/src/nes/nes_rom.c#L163) and NOT `NOFRENDO_MALLOC()`.

Same for `vrom`.

TL;DR: `NOFRENDO_MALLOC()`/`NOFRENDO_FREE()` operate with pointers to special data structures while `malloc()`/`free()` work with allocated memory directly.

**Comment by moononournation** [comment]
It is odd for this inconsistent:
```C
    if ((*rominfo)->sram)
       NOFRENDO_FREE((*rominfo)->sram);
    if ((*rominfo)->rom)
       free((*rominfo)->rom);
    if ((*rominfo)->vrom)
       free((*rominfo)->vrom);
    if ((*rominfo)->vram)
       NOFRENDO_FREE((*rominfo)->vram);
```

Is it possible alter malloc() instead?

**Comment by and3rson** [comment]
@moononournation We could replace all malloc/free calls with NOFRENDO_*, but maybe there's some reason why they are inconsistent (just a guess). This PR just fixes a bug caused by incorrect mixing of those functions.
